### PR TITLE
Switch to Twill 0.8.0-SNAPSHOT

### PR DESCRIPTION
**Pull request by gokulavasan**
Change code to use the new Discoverable class

Build : http://builds.cask.co/browse/CDAP-RUT167

### CODE REVIEW

File: cdap-app-fabric/src/main/java/org/apache/tephra/inmemory/InMemoryTransactionService.java [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+import org.apache.twill.common.Cancellable;
+import org.apache.twill.discovery.Discoverable;
+import org.apache.twill.discovery.DiscoveryService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.net.InetSocketAddress;
+
+/**
+ * Note: Had to copy this class since Tephra depends on old Twill release where Discoverable is an interface. We can
```
Comment (poornachandra) [inline cdap-app-fabric/src/main/java/org/apache/tephra/inmemory/InMemoryTransactionService.java]: Can you file a Tephra JIRA for this? Also file a shadow CDAP JIRA and put the JIRA id here.

**Comment by gokulavasan** [comment]
@poornachandra Addressed comment. Please take a look again.

File: cdap-app-fabric/src/main/java/org/apache/twill/yarn/YarnTwillRunnerService.java [hunk 2]
```diff
@@ -322,7 +321,6 @@ public synchronized TwillController lookup(String applicationName, final RunId r
   }
 
   private void startUp() throws Exception {
-    yarnAppClient.startAndWait();
```
Comment (poornachandra) [inline cdap-app-fabric/src/main/java/org/apache/twill/yarn/YarnTwillRunnerService.java:325]: Is this not required anymore?

File: cdap-common/src/main/java/org/apache/twill/filesystem/FileContextLocation.java [hunk 3]
```diff
@@ -80,6 +80,32 @@ public boolean createNew() throws IOException {
   }
 
   @Override
+  public boolean createNew(String permission) throws IOException {
```
Comment (poornachandra) [inline cdap-common/src/main/java/org/apache/twill/filesystem/FileContextLocation.java]: I see that Twill version is upgraded in the pom, why do we still need to copy this class?

File: cdap-client-tests/pom.xml [hunk 4]
```diff
@@ -63,6 +63,12 @@ the License.
     </dependency>
     <dependency>
       <groupId>co.cask.cdap</groupId>
+      <artifactId>cdap-data-fabric</artifactId>
```
Comment (poornachandra) [inline cdap-client-tests/pom.xml:66]: Why is this required?

**Review by poornachandra — COMMENTED** [review]
@gokulavasan Changes mostly look good. Just a few questions.

File: cdap-app-fabric/src/main/java/org/apache/twill/yarn/YarnTwillRunnerService.java (hunk 2 above)
Comment (gokulavasan) [reply cdap-app-fabric/src/main/java/org/apache/twill/yarn/YarnTwillRunnerService.java:325]: Yes it is not required. YarnAppClient is not a service anymore in Twill 0.8.0. @anwar6953 Can you confirm?

File: cdap-common/src/main/java/org/apache/twill/filesystem/FileContextLocation.java (hunk 3 above)
Comment (gokulavasan) [reply cdap-common/src/main/java/org/apache/twill/filesystem/FileContextLocation.java]: We added new methods to the Location interface. I copied the implementation from Twill. Can you confirm if this is OK @anwar6953 ?

File: cdap-client-tests/pom.xml (hunk 4 above)
Comment (gokulavasan) [reply cdap-client-tests/pom.xml:66]: This is required so that copied Tephra classes take precedence over tephra jars. This is required to make the unit tests pass.

File: cdap-app-fabric/src/main/java/org/apache/twill/yarn/YarnTwillRunnerService.java (hunk 2 above)
Comment (anwar6953) [reply cdap-app-fabric/src/main/java/org/apache/twill/yarn/YarnTwillRunnerService.java:325]: We made it no longer a Service, since we don't need to do anything in startup of it anymore.

File: cdap-common/src/main/java/org/apache/twill/filesystem/FileContextLocation.java (hunk 3 above)
Comment (anwar6953) [inline cdap-common/src/main/java/org/apache/twill/filesystem/FileContextLocation.java]: According to the comment in https://issues.cask.co/browse/CDAP-5328, we shouldn't need to copy the `FileContextLocation` class anymore.
Check if there's any changes in cdap that don't exist in twill. If nothing, we can probably just remove this class from CDAP.

Comment (anwar6953) [reply cdap-common/src/main/java/org/apache/twill/filesystem/FileContextLocation.java]: According to the comment in https://issues.cask.co/browse/CDAP-5328, we shouldn't need to copy the `FileContextLocation` class anymore.
Check if there's any changes in cdap that don't exist in twill. If nothing, we can probably just remove this class from CDAP.

File: cdap-app-fabric/src/main/java/org/apache/twill/yarn/YarnTwillRunnerService.java (hunk 2 above)
Comment (poornachandra) [reply cdap-app-fabric/src/main/java/org/apache/twill/yarn/YarnTwillRunnerService.java:325]: 👍

File: cdap-client-tests/pom.xml (hunk 4 above)
Comment (poornachandra) [reply cdap-client-tests/pom.xml:66]: I see. Would it be possible to move all tephra changes into a single commit, so that it is easy to revert later?

File: cdap-common/src/main/java/org/apache/twill/filesystem/FileContextLocation.java (hunk 3 above)
Comment (poornachandra) [reply cdap-common/src/main/java/org/apache/twill/filesystem/FileContextLocation.java]: @gokulavasan Maybe other classes that are copied over from Twill can also be removed then. Can you check?

File: cdap-client-tests/pom.xml (hunk 4 above)
Comment (gokulavasan) [reply cdap-client-tests/pom.xml:66]: I have it spread across multiple commits. I will try to move them in a separate single commit if it is definitely required. Otherwise, I will add enough to the JIRA to make it easy to remove later.

File: cdap-common/src/main/java/org/apache/twill/filesystem/FileContextLocation.java (hunk 3 above)
Comment (gokulavasan) [reply cdap-common/src/main/java/org/apache/twill/filesystem/FileContextLocation.java]: I checked the JIRA. CDAP-5328, the other class that is used is FileContextLocationFactory but that class has changes in CDAP that is not present in Twill. How to find out what classes to remove in CDAP?

File: pom.xml [hunk 5]
```diff
@@ -135,7 +139,7 @@
     <slf4j.version>1.7.5</slf4j.version>
     <tephra.version>0.8.0-incubating</tephra.version>
     <thrift.version>0.9.3</thrift.version>
-    <twill.version>0.7.0-incubating</twill.version>
+    <twill.version>0.8.0-SNAPSHOT</twill.version>
```
Comment (poornachandra) [inline pom.xml:142]: Should we add a JIRA to remove snapshot before releasing 3.6?

File: cdap-client-tests/pom.xml (hunk 4 above)
Comment (poornachandra) [reply cdap-client-tests/pom.xml:66]: Okay, let's document what classes to remove in the JIRA

**Review by poornachandra — APPROVED** [review]
@gokulavasan Changes LGTM, a couple of comments.

File: pom.xml (hunk 5 above)
Comment (gokulavasan) [reply pom.xml:142]: Sure, will create one now.

**Comment by gokulavasan** [comment]
Thanks for the review @poornachandra. Created https://issues.cask.co/browse/CDAP-7277 to track the removal of SNAPSHOT.

**Comment by gokulavasan** [comment]
Build passed. Merging.
